### PR TITLE
feat: `compute_note_hash_and_nullifier` check

### PR DESCRIPTION
**Pull request by benesjan**
# Description

Checks whether Aztec contract has `compute_note_hash_and_nullifier` function defined when `Storage` struct is present.

Resolves https://github.com/AztecProtocol/aztec-packages/issues/2738

# PR Checklist\*

- [ ] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/aztec_library.rs [hunk 1]
```diff
@@ -236,9 +242,20 @@ fn is_custom_attribute(attr: &SecondaryAttribute, attribute_name: &str) -> bool
...
-fn transform_module(module: &mut SortedModule, storage_defined: bool) -> bool {
+fn transform_module(module: &mut SortedModule) -> Result<bool, (DefCollectorErrorKind, FileId)> {
     let mut has_transformed_module = false;
 
+    // Check for a user defined storage struct
+    let storage_defined = check_for_storage_definition(&module);
+
+    if storage_defined && check_for_compute_note_hash_and_nullifier_definition(&module) {
+        // TODO: do a refactor so that file_id is real --> you want to get crate_graph.root_file_id
+        return Err((
```
[the remark this answers is not in the document]
Comment (benesjan) [reply compiler/noirc_frontend/src/hir/aztec_library.rs:256]: Addressed in https://github.com/noir-lang/noir/pull/3216/commits/fef1ac5d480d215526db9e9cccb28469e697a752

File: compiler/noirc_frontend/src/hir/aztec_library.rs [hunk 2]
```diff
@@ -209,19 +204,30 @@ fn include_relevant_imports(ast: &mut SortedModule) {
...
-fn check_for_aztec_dependency(crate_id: &CrateId, context: &Context) -> Result<(), FileId> {
+fn check_for_aztec_dependency(
+    crate_id: &CrateId,
+    context: &Context,
+) -> Result<bool, (DefCollectorErrorKind, FileId)> {
     let crate_graph = &context.crate_graph[crate_id];
     let has_aztec_dependency = crate_graph.dependencies.iter().any(|dep| dep.as_name() == "aztec");
     if has_aztec_dependency {
-        Ok(())
+        Ok(true)
```
[the remark this answers is not in the document]
Comment (benesjan) [reply compiler/noirc_frontend/src/hir/aztec_library.rs]: Addressed in https://github.com/noir-lang/noir/pull/3216/commits/be07967c8106884d4ce154ab6ca5f7185c21da55

**Review by LHerskind — APPROVED** [review]
Looks good to me, but I'm an idiot :shrug:.

File: compiler/noirc_frontend/src/hir/aztec_library.rs [hunk 3]
```diff
@@ -209,19 +203,30 @@ fn include_relevant_imports(ast: &mut SortedModule) {
...
-    module.types.iter().any(|function| function.name.0.contents == "Storage")
+    module.types.iter().any(|r#struct| r#struct.name.0.contents == "Storage")
+}
+
+// Check to see if the user has defined a compute_note_hash_and_nullifier function exists
+fn check_for_compute_note_hash_and_nullifier_definition(module: &SortedModule) -> bool {
+    module
+        .functions
+        .iter()
+        .any(|func| func.def.name.0.contents == "compute_note_hash_and_nullifier")
```
Comment (Maddiaa0) [inline compiler/noirc_frontend/src/hir/aztec_library.rs:229]: now i think of it, i think that compute note hash and nullifier will also cause errors if the function selector is not identical to the one specified in the docs. 
I do not want to block this pr, but we probably need:
    unconstrained fn compute_note_hash_and_nullifier(contract_address: Field, nonce: Field, storage_slot: Field, preimage: [Field; TOKEN_NOTE_LEN]) -> [Field; 4] 

to check for the types

**Review by Maddiaa0 — COMMENTED** [review]
code looks good, have a possible follow up issue

**Review by Maddiaa0 — APPROVED** [review]
nice lgtm

**Pull request merged**